File: Code/CH8 Code/VB/Vlad-vb/Vlad.vb
Option Strict On

Imports System
Imports System.Collections
Imports System.Windows.Forms
Imports System.ComponentModel
Imports System.Text.RegularExpressions

<ProvideProperty("Required", GetType(Control)), _
 ProvideProperty("RangeValueRequired", GetType(Control)), _
 ProvideProperty("MinValue", GetType(Control)), _
 ProvideProperty("MaxValue", GetType(Control))> _
Public Class NumberValidate
  Inherits Component
  Implements IExtenderProvider

#Region "locals"

  ' holds Key - value pair for efficient retrieval
  'Holds all properties of all controls that use this extender.
  Private CustomProps As System.Collections.Hashtable = _
                        New System.Collections.Hashtable()

  Private er As ErrorProvider = New ErrorProvider()
  Private tp As ToolTip = New ToolTip()

  'Holds all the custom properties of a control
  Private Class Props
    Public Required As Boolean = False
    Public RangeRequired As Boolean = False
    Public MinVal As Decimal = 0
    Public MaxVal As Decimal = 999
  End Class

#End Region

  Public Sub New()

  End Sub

  Function CanExtend(ByVal target As Object) As Boolean _
                     Implements IExtenderProvider.CanExtend
    If TypeOf target Is TextBox Then
      Dim t As TextBox = CType(target, TextBox)
      tp.Active = True
      Return True
    Else
      Return False
    End If
  End Function

  Private Sub SetToolTip(ByVal ctl As Control)
    Dim tip As String = String.Empty
    Dim p As Props = CType(CustomProps(ctl), Props)
    If p.Required Then
      tip = "Validation of numbers required"
    End If
    If p.RangeRequired Then
      tip += " / Number range required"
      tip += " / Min value = " + p.MinVal.ToString()
      tip += " / Max value = " + p.MaxVal.ToString()
    End If
    tp.SetToolTip(ctl, tip)
  End Sub


#Region "RangeValue required"

  'Get method for range required
  Public Function GetRangeValueRequired(ByVal ctl As Control) As Boolean
    'This makes best use of a hashtable for quick retrieval
    If CustomProps.Contains(ctl) Then
      Dim p As Props = CType(CustomProps(ctl), Props)
      SetToolTip(ctl)
      Return p.RangeRequired
    End If
    Return False
  End Function

  'Set method for Range required
  Public Sub SetRangeValueRequired(ByVal ctl As Control, ByVal val As Boolean)
    If CustomProps.Contains(ctl) Then
      'See if this property is already correctly set
      Dim p As Props = CType(CustomProps(ctl), Props)
      If val = p.RangeRequired Then
        Return
      End If

      'Set this property and add it back to the list
      p.RangeRequired = val
      CustomProps(ctl) = p
      If val Then
        AddHandler ctl.KeyPress, _
                  New KeyPressEventHandler(AddressOf KeyPress)
        AddHandler ctl.Validating, _
                  New CancelEventHandler(AddressOf ValidateProp)
      Else
        RemoveHandler ctl.KeyPress, _
                  New KeyPressEventHandler(AddressOf KeyPress)
        RemoveHandler ctl.Validating, _
                  New CancelEventHandler(AddressOf ValidateProp)
      End If
      SetToolTip(ctl)
    Else
      'Set this property and add it to the list
      Dim p As Props = New Props()
      p.RangeRequired = val
      CustomProps.Add(ctl, p)
      If val Then
        AddHandler ctl.KeyPress, New KeyPressEventHandler(AddressOf KeyPress)
      End If
      SetToolTip(ctl)
    End If
  End Sub

#End Region

#Region "Min and Max Values"

  Public Function GetMinValue(ByVal ctl As Control) As Decimal
    If CustomProps.Contains(ctl) Then
      Dim p As Props = CType(CustomProps(ctl), Props)
      SetToolTip(ctl)
      Return p.MinVal
    End If
    Return 0
  End Function

  '   <DefaultValue(0)>_
  Public Sub SetMinValue(ByVal ctl As Control, ByVal val As Decimal)
    If val < 0 Then val = 0

    If CustomProps.Contains(ctl) Then
      Dim p As Props = CType(CustomProps(ctl), Props)
      p.MinVal = val
      CustomProps(ctl) = p
      SetToolTip(ctl)
    Else
      Dim p As Props = New Props()
      p.MinVal = val
      CustomProps.Add(ctl, p)
      SetToolTip(ctl)
    End If
  End Sub

  Public Function GetMaxValue(ByVal ctl As Control) As Decimal
    If (CustomProps.Contains(ctl)) Then
      Dim p As Props = CType(CustomProps(ctl), Props)
      SetToolTip(ctl)
      Return p.MaxVal
    End If
    Return 999
  End Function

  ' [DefaultValue(999)]
  Public Sub SetMaxValue(ByVal ctl As Control, ByVal val As Decimal)
    If val > 999 Then val = 999

    If CustomProps.Contains(ctl) Then
      Dim p As Props = CType(CustomProps(ctl), Props)
      p.MaxVal = val
      CustomProps(ctl) = p
      SetToolTip(ctl)
    Else
      Dim p As Props = New Props()
      p.MaxVal = val
      CustomProps.Add(ctl, p)
      SetToolTip(ctl)
    End If
  End Sub

#End Region

#Region "Required property"

  Public Function GetRequired(ByVal ctl As Control) As Boolean
    If CustomProps.Contains(ctl) Then
      Dim p As Props = CType(CustomProps(ctl), Props)
      SetToolTip(ctl)
      Return p.Required
    End If
    Return False
  End Function

  Public Sub SetRequired(ByVal ctl As Control, ByVal val As Boolean)
    If CustomProps.Contains(ctl) Then
      Dim p As Props = CType(CustomProps(ctl), Props)
      If val = p.Required Then
        Return
      End If

      p.Required = val
      CustomProps(ctl) = p
      SetToolTip(ctl)
      If val Then
        AddHandler ctl.KeyPress, _
                  New KeyPressEventHandler(AddressOf KeyPress)
        AddHandler ctl.Validating, _
                  New CancelEventHandler(AddressOf ValidateProp)
      Else
        RemoveHandler ctl.KeyPress, _
                  New KeyPressEventHandler(AddressOf KeyPress)
        RemoveHandler ctl.Validating, _
                  New CancelEventHandler(AddressOf ValidateProp)
      End If
    Else
      Dim p As Props = New Props()
      p.Required = val

      CustomProps.Add(ctl, p)
      SetToolTip(ctl)
      AddHandler ctl.Validating, New CancelEventHandler(AddressOf ValidateProp)
    End If
  End Sub

#End Region

#Region "events"

  Private Sub KeyPress(ByVal sender As Object, ByVal e As KeyPressEventArgs)
    'Allow backspace
    If e.KeyChar = Chr(8) Then
      Return
    End If

    'Allow 0-9
    If Not Regex.IsMatch(e.KeyChar.ToString(), "[0-9]") Then
      e.Handled = True
    End If
  End Sub

  Private Sub ValidateProp(ByVal sender As Object, ByVal e As CancelEventArgs)
    Dim ctl As Control = CType(sender, Control)

    'Reset the error
    er.SetError(ctl, "")

    If GetRequired(ctl) Then
      If ctl.Text = String.Empty Then
        er.SetError(ctl, "No value was entered when one was required")
        Return
      End If

      If GetRangeValueRequired(ctl) Then
        Dim p As Props = CType(CustomProps(ctl), Props)
        If Decimal.Parse(ctl.Text) < p.MinVal Then
          er.SetError(ctl, "Value entered is less than minimum value")
        End If
        If Decimal.Parse(ctl.Text) > p.MaxVal Then
          er.SetError(ctl, "Value entered is greater than minimum value")
        End If
      End If
    End If
  End Sub

#End Region

End Class

File: Code/CH8 Code/C#/Vlad_c/Vlad.cs
using System;
using System.Collections;
using System.Windows.Forms;
using System.ComponentModel;
using System.Text.RegularExpressions;

namespace Vlad_C
{

  [ProvideProperty("Required",            typeof(Control))]
  [ProvideProperty("RangeValueRequired",  typeof(Control))]
  [ProvideProperty("MinValue",            typeof(Control))]
  [ProvideProperty("MaxValue",            typeof(Control))]
  public class NumberValidate : Component, IExtenderProvider
  {
    #region locals

    // holds Key - value pair for efficient retrieval
    //Holds all properties of all controls that use this extender.
    private System.Collections.Hashtable CustomProps  = 
                        new System.Collections.Hashtable();

    ErrorProvider er = new ErrorProvider();
    ToolTip       tp = new ToolTip();
    
    //Holds all the custom properties of a control
    private class Props
    {
      public bool Required      = false;
      public bool RangeRequired = false;
      public Decimal  MinVal    = 0;
      public Decimal  MaxVal    = 999;
    }

    #endregion


    public NumberValidate()
    {
    }

    bool IExtenderProvider.CanExtend(object target) 
    {
      if (target is TextBox) 
      {
        TextBox t = (TextBox)target;
        tp.Active = true;
        return true;
      }
      else
        return false;
    }

    private void SetToolTip(Control ctl)
    {
      string tip = string.Empty;
      Props p = (Props)CustomProps[ctl];
      if(p.Required)
        tip = "Validation of numbers required";
      if(p.RangeRequired)
      {
        tip += " / Number range required";
        tip += " / Min value = " + p.MinVal.ToString();
        tip += " / Max value = " + p.MaxVal.ToString();
      }
      tp.SetToolTip(ctl, tip);
    }

    #region RangeValue required

    //Get method for range required
    public bool GetRangeValueRequired(Control ctl)
    {
      //This makes best use of a hashtable for quick retrieval
      if(CustomProps.Contains(ctl))
      {
        Props p = (Props)CustomProps[ctl];
        SetToolTip(ctl);
        return p.RangeRequired;
      }
      else
        return false;
    }

    //Set method for Range required
    public void SetRangeValueRequired(Control ctl, bool val)
    {
      if(CustomProps.Contains(ctl))
      {
        //See if this property is already correctly set
        Props p = (Props)CustomProps[ctl];
        if(val == p.RangeRequired)
          return;

        //Set this property and add it back to the list
        p.RangeRequired = val;
        CustomProps[ctl] = p;
        if(val)
        {
          ctl.KeyPress   += new KeyPressEventHandler(KeyPress);
          ctl.Validating += new CancelEventHandler(ValidateProp);
        }
        else
        {
          ctl.KeyPress   -= new KeyPressEventHandler(KeyPress);
          ctl.Validating -= new CancelEventHandler(ValidateProp);
        }
        SetToolTip(ctl);
      }
      else
      {
        //Set this property and add it to the list
        Props p = new Props();
        p.RangeRequired = val;
        CustomProps.Add(ctl, p);
        if(val)
          ctl.KeyPress += new KeyPressEventHandler(KeyPress);

        SetToolTip(ctl);
      }
    }

    #endregion

    #region Min and Max Values

    public Decimal GetMinValue(Control ctl)
    {
      if(CustomProps.Contains(ctl))
      {
        Props p = (Props)CustomProps[ctl];
        SetToolTip(ctl);
        return p.MinVal;
      }
      else
        return 0;
    }

    [DefaultValue(0)]
    public void SetMinValue(Control ctl, Decimal val)
    {
      if(val < 0)
        val = 0;

      if(CustomProps.Contains(ctl))
      {
        Props p = (Props)CustomProps[ctl];
        p.MinVal = val;
        CustomProps[ctl] = p;
        SetToolTip(ctl);
      }
      else
      {
        Props p = new Props();
        p.MinVal = val;
        CustomProps.Add(ctl, p);
        SetToolTip(ctl);
      }
    }

    public Decimal GetMaxValue(Control ctl)
    {
      if(CustomProps.Contains(ctl))
      {
        Props p = (Props)CustomProps[ctl];
        SetToolTip(ctl);
        return p.MaxVal;
      }
      else
        return 999;
    }

    [DefaultValue(999)]
    public void SetMaxValue(Control ctl, Decimal val)
    {
      if(val > 999)
        val = 999;

      if(CustomProps.Contains(ctl))
      {
        Props p = (Props)CustomProps[ctl];
        p.MaxVal = val;
        CustomProps[ctl] = p;
        SetToolTip(ctl);
      }
      else
      {
        Props p = new Props();
        p.MaxVal = val;
        CustomProps.Add(ctl, p);
        SetToolTip(ctl);
      }
    }

    #endregion

    #region Required property

    public bool GetRequired(Control ctl)
    {
      if(CustomProps.Contains(ctl))
      {
        Props p = (Props)CustomProps[ctl];
        SetToolTip(ctl);
        return p.Required;
      }
      else
        return false;
    }

    public void SetRequired(Control ctl, bool val)
    {
      if(CustomProps.Contains(ctl))
      {
        Props p = (Props)CustomProps[ctl];
        if(val == p.Required)
          return;

        p.Required = val;
        CustomProps[ctl] = p;
        SetToolTip(ctl);
        if(val)
        {
          ctl.KeyPress   += new KeyPressEventHandler(KeyPress);
          ctl.Validating += new CancelEventHandler(ValidateProp);
        }
        else
        {
          ctl.KeyPress   -= new KeyPressEventHandler(KeyPress);
          ctl.Validating -= new CancelEventHandler(ValidateProp);
        }
      }
      else
      {
        Props p = new Props();
        p.Required = val;

        CustomProps.Add(ctl, p);
        SetToolTip(ctl);
        ctl.Validating += new CancelEventHandler(ValidateProp);
      }
    }

    #endregion

    #region events

    private void KeyPress(object sender, KeyPressEventArgs e)
    {
      //Allow backspace
      if(e.KeyChar == (char)8)
        return;

      //Allow 0-9
      if(!Regex.IsMatch(e.KeyChar.ToString(), "[0-9]"))
        e.Handled = true;
    }

    private void ValidateProp(object sender, CancelEventArgs e)
    {
      Control ctl = (Control)sender;

      //Reset the error
      er.SetError(ctl, "");

      if(GetRequired(ctl))
      {
        if(ctl.Text == string.Empty)
        {
          er.SetError(ctl, "No value was entered when one was required");
          return;
        }

        if(GetRangeValueRequired(ctl))
        {
          Props p = (Props)CustomProps[ctl];
          if(Decimal.Parse(ctl.Text) < p.MinVal)
            er.SetError(ctl, "Value entered is less than minimum value");
          if(Decimal.Parse(ctl.Text) > p.MaxVal)
            er.SetError(ctl, "Value entered is greater than minimum value");
        }
      }
    }

    #endregion
  }

}
